[RLlib; offline RL; docs] Docs do-over (new API stack): Offline RL.

simonsays1980 opened the pull request:
## Why are these changes needed?

With the new Offline RL API soon to be released the corresponding docs need an overhaul. This PR includes all these changes with:

- An introduction to the new API and its key principles
- Examples of how to record expert data, train on expert data, convert external data and old `SampleBatch` data.
- Hints and notes about data formats, underlying procedures, etc.
- Scaling and tuning guides (to come) and descriptions about internals

These docs should enable old and new RLlib users to implement offline RL on their data fast and smoothly. Furthermore, it should apply to real industry use cases which need to digest peta-bytes of offline data to train policies for highly-complex tasks.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

simonsays1980 committed:
Started rewriting Offline RL documentation and added examples, descriptions, hints, notes, and testcode.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

sven1977 commented:
One warning, which breaks the readthedocs build. I think it's this one:
```
/home/docs/checkouts/readthedocs.org/user_builds/anyscale-ray/checkouts/48818/doc/source/rllib/rllib-offline.rst:666: WARNING: Enumerated list ends without a blank line; unexpected unindent.
```

simonsays1980 committed:
Fixed indent error.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Removed :hide: syntax from testsetup b/c it was producing an error.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+
+Example: Training an expert policy.
+--------------------------------------------
+In this example you train an PPO agent to play ``CartPole-v1`` until it collects an episode mean reward of ``450.0``. You checkpoint 
```
```suggestion
In this example you train a PPO agent on the ``CartPole-v1`` environment until it reaches an episode mean return of ``450.0``. You checkpoint 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    # the tabular recording data.
+    tabular_data_path = None
+
+.. testcode::
```
remove this line?

simonsays1980 replied:
Nope, the `testsetup` is hidden for users. We define there only some global variables we can use across all examples :)

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    config = (
+        PPOConfig()
+        .environment("CartPole-v1")
+        # Enable the new API stack.
+        .api_stack(
+            enable_rl_module_and_learner=True,
+            enable_env_runner_and_connector_v2=True,
+        )
```
```suggestion
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+            ),
+        ),
+    )
+    analysis = tuner.fit()
```
rename to `results` (name used across RLlib).

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    analysis = tuner.fit()
 
-    For custom models and enviroments, you'll need to use the `Python API <rllib-training.html#basic-python-api>`__.
+    best_checkpoint = (
```
nit: add some comments here about what we are going to do with this checkpoint next ...

simonsays1980 replied:
Yeah! Good point!

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    )
 
-In this example, we will save batches of experiences generated during online training to disk, and then leverage this saved data to train a policy offline using DQN. First, we run a simple policy gradient algorithm for 100k steps with ``"output": "/tmp/cartpole-out"`` to tell RLlib to write simulation outputs to the ``/tmp/cartpole-out`` directory.
+Example: Record expert data to local disk
```
Maybe add a 1-2 sentences here between the paragraphs about that we are going to do next with the generated checkpoint.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+        .environment(
+            env="CartPole-v1",
+        )
+        # Make sure you use the new API stack.
+        .api_stack(
+            enable_rl_module_and_learner=True,
+            enable_env_runner_and_connector_v2=True,
+        )
```
```suggestion
```

simonsays1980 replied:
Finally, it is default :)

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    algo = config.build()
+    # Load now the PPO-trained `RLModule` to use in recording.
+    algo.restore_from_path(
+        best_checkpoint,
```
I think you have to do this here:
```
best_checkpoint / COMPONENT_LEARNER_GROUP / COMPONENT_LEARNER / COMPONENT_RL_MODULE
```
... to make this work ...

simonsays1980 replied:
Really? It worked like this in my example locally ...

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+        # Sample 50 episodes in each evaluation rollout.
+        .evaluation(
+            evaluation_num_env_runners=5,
+            evaluation_duration=50,
```
let's also specify the unit here, just for clarity `evaluation_duration_unit="episodes"`

simonsays1980 replied:
Alright, implemented it! Thanks for the suggestion!

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
 
-In practice, when training on offline data, it is usually not straightforward to evaluate the trained policies using a simulator as in online RL. For example, in recommender systems, rolling out a policy trained on offline data in a real-world environment can jeopardize your business if the policy is suboptimal. For these situations we can use `off-policy estimation <https://arxiv.org/abs/1911.06 ...
+.. hint:: RLlib stores records under a folder named by the RL environment. Therein, you see for each ``OfflineSingleAgentEnvRunner`` and write operation 
+    one folder of Parquet files. The write operation count is given in the second numbering. 
```
so this is the 6, 11, 16, 21? We should explain what this means and why exactly we get these numbers.

simonsays1980 replied:
I have described in more detailed how these numbers occur. I will also in another PR bring in some changes to the `OfflineSingleAgentEnvRunner`.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+                Columns.REWARDS: "r_t",
+                Columns.NEXT_OBS: "o_tp1",
+                Columns.INFOS: "i_t",
+                "done": "d_t",
```
Use `Columns.DONES`?

Or should we even use `TERMINATEDS` here? (not sure what effects using "dones" would have on the algo loss logic).

simonsays1980 replied:
We don't have `Columns.DONES` :) I can create them.

sven1977 replied:
Nope, that's fine! Let's leave it as-is with this comment here.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+        )
+    )
+        
+.. note:: Internally, the legacy ``gym``'s ``done`` signals are mapped to ``gymnasium``'s ``terminated`` signals, with ``truncated`` values defaulting to 
```
Ah, ok, perfect! Forget my comment above :)

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
-.. hint:: You should use separate datasets for algorithm training and OPE, as shown here.
+.. note::
+    RLlib processes tabular data in batches, converting each row into a *single-step episode*. This approach is primarily for procedural simplicity, as data cannot 
+    generally be assumed to arrive in time-ordered rows grouped by episodes, though this may occasionally be the case (however knowledge of such a structure resides  
```
problem very nicely described!

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    with the user as RLlib cannot easily infer it automatically). While it is possible to concatenate ``SingleAgentEpisode`` chunks, this cannot be done with fragmented chunks.
 
-We can now train a DQN algorithm offline and evaluate it using OPE:
+If you require full trajectories you can transform your tabular data into ``SingleAgentEpisode`` objects and store these in Parquet format. The next example shows how to do this. 
```
In general across this rst, always use the:
```
:py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
```

notation for classes and the:

```
:py:meth:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode.get_observations`
```

notation for individual class methods.

This automatically creates links into these APIs' reference pages.

sven1977 replied:
So replace all ``SingleAgentEpisode`` with
```
:py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
```

etc..

simonsays1980 replied:
@sven1977 Some of these classes and their methods appear not to be in the docs.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+        .environment(
+            env="CartPole-v1",
+        )
+        # Make sure you use the new API stack.
+        .api_stack(
+            enable_rl_module_and_learner=True,
+            enable_env_runner_and_connector_v2=True,
+        )
```
```suggestion
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
-        algo.train()
+    # Load now the PPO-trained `RLModule` to use in recording.
+    algo.restore_from_path(
+        best_checkpoint,
```
same here, correct?

We need to specify the individual RLModule sub-path instead of this algo top-level path.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+.. note::
+    ``infos`` are not stored to disk when they are all empty.
+
+If your expert data is given in columnar format and you need to train on full expert trajectories you can follow the code in the following example to convert 
```
These examples are all really useful!

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -250,20 +695,7 @@ To load sample data using Dataset, specify input and input_config keys like the
-        }
-    }
 
 Writing Environment Data
```
Dumb question: Are these paragraphs still needed (new API stack)?
If not (or if they only refer to old API stack stuff), let's completely remove them.

sven1977 replied:
Same question for the above `Ray Data Integration` section.

simonsays1980 replied:
Yes they are all old API stack and will be removed by my soon. They helped me structuring the file better.

simonsays1980 committed:
Added more sections to new offline docs. In addition completed some missing docstrings in 'AlgorithmConfig'.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added suggestions from @sven1977's review and some clean-ups.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Removed some more enablements of the new API stack. This is not needed becvause the new API stack is now the default stack.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added missing docstring for 'output_write_episodes' and added section 'Output API' to docs.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Started writing offline scaling guide.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added documentation for performance tuning in Offline RL. Performance tuning of data read operations with available resources, data locality, data sharding, and data pruning.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed some small nits.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added sections for tuning post-processing and training in RLlib's Offline RL API with multiple tuning tips.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added sections for tuning training with batch sizes, prefetching and substream iterations. Fixed a couple of reference bugs.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Finished 'How to tune performance' section and started with 'Customization' section.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added extensive example for overriding 'OfflinePreLearner'.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Finished customization section.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added section 'Monitoring' and made bug fixes.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

sven1977 commented on doc/source/rllib/rllib-rlmodule.rst:
```diff
@@ -228,7 +228,7 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
 
 
-Also the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` class's constrcutor requires a dataclass config object called `~ray.rllib.core.rl_module.rl_module.RLModuleConfig` which contains the following fields:
+Also the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` class's constructor requires a dataclass config object called :py:class:`~ray.rllib.core.rl_module.rl_module.RLModuleConfig` which contains the following fields:
```
Nice!

angelinalg replied:
```suggestion
Also the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` class's constructor requires a dataclass config object called :py:class:`~ray.rllib.core.rl_module.rl_module.RLModuleConfig`, which contains the following fields:
```

sven1977 committed:
Run vale CLI

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+
+RLlib represents trajectory sequences (for example, ``(s, a, r, s', ...)`` tuples) with :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
+objects (multi-agent offline training is currently not supported). Using this episode format 
+enables efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getters.
```
```suggestion
allows for efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getters.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+objects (multi-agent offline training is currently not supported). Using this episode format 
+enables efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getters.
+During online training, RLlib uses :py:class:`~ray.rllib.env.single_agent_env_runner.SingleAgentEnvRunner` actors to generate episodes of experiences 
+in parallel using the current policy. RLlib can also uses this same episode format for reading and writing experiences to offline storage. For offline 
```
```suggestion
in parallel using the current policy. However, RLlib uses this same episode format for reading experiences from and writing experiences to offline storage. For offline 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+enables efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getters.
+During online training, RLlib uses :py:class:`~ray.rllib.env.single_agent_env_runner.SingleAgentEnvRunner` actors to generate episodes of experiences 
+in parallel using the current policy. RLlib can also uses this same episode format for reading and writing experiences to offline storage. For offline 
+writing RLlib uses the :py:class:`~ray.rllib.offline.offline_env_runner.OfflineSingleAgentEnvRunner` that samples episodes and writes them to offline 
+storage. 
```
```suggestion
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+#. You need to read the data easily with other data tools or ML libraries.
+#. You don't need full (usually complete) trajectories in training.
+
+.. note:: RLlib's redesigned API stack incorporates principles that support standalone applications. Consequently, the 
```
```suggestion
.. note:: RLlib's new API stack incorporates principles that support standalone applications. Consequently, the 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+Most importantly, RLlib's offline RL API builds on top of :ref:`Ray Data <data>` and therefore features in general all read and 
+write methods supported by Ray Data (for example :py:class:`~ray.data.read_parquet`, :py:class:`~ray.data.read_json`, etc.) with
+:py:class:`~ray.data.read_parquet` and :py:class:`~ray.data.Dataset.write_parquet` being the default read and write methods. A core design principle 
+of the API is to apply as many data transformations as possible on-the-fly to the learner, allowing the learner to focus exclusively on model updates. 
```
```suggestion
of the API is to apply as many data transformations as possible on-the-fly prior to engaging the learner, allowing the latter to focus exclusively on model updates. 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+:py:class:`~ray.data.read_parquet` and :py:class:`~ray.data.Dataset.write_parquet` being the default read and write methods. A core design principle 
+of the API is to apply as many data transformations as possible on-the-fly to the learner, allowing the learner to focus exclusively on model updates. 
+
+.. hint:: During the transition phase from old to new API stack you can use the new offline RL API also with your 
```
```suggestion
.. hint:: During the transition phase from old- to new API stack you can use the new offline RL API also with your 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+    # Run 10 evaluation iterations and record the data.
+    for i in range(10):
+        print(f"Iteration {i + 1}")
+        res_eval = algo.evaluate()
+        print(res_eval)
```
```suggestion
        eval_results = algo.evaluate()
        print(eval_results)
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
-With RLlib's evaluation framework you can:
 
-- Evaluate policies on a simulated environment, if available, using ``evaluation_config["input"] = "sampler"``. You can then monitor your policy's performance on tensorboard as it is getting trained (by using ``tensorboard --logdir=~/ray_results``).
+.. hint:: RLlib stores records under a folder named by the RL environment. Therein, you see for each :py:class:`~ray.rllib.offline.offline_env_runner.OfflineSingleAgentEnvRunner` 
+    and write operation one folder of Parquet files. The write operation count is given in the second numbering. For example: above, env-runner 1 has sampled 25 episodes at 
```
```suggestion
.. hint:: RLlib stores records under a folder named by the RL environment. Therein, you see one folder of Parquet files for each :py:class:`~ray.rllib.offline.offline_env_runner.OfflineSingleAgentEnvRunner` 
    and write operation. The write operation count is given in the second numbering. For example: above, env-runner 1 has sampled 16 episodes at 
```

sven1977 replied:
Is the 25 correct here? The filename is 00016, no?

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+        ...
+        .offline_data(
+            input_=[<input_path>],
+            # Provide the schema of your data.
```
```suggestion
            # Provide the schema of your data (map to column names known to RLlib).
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+.. note::
+    RLlib processes tabular data in batches, converting each row into a *single-step episode*. This approach is primarily for procedural simplicity, as data can't
+    generally be assumed to arrive in time-ordered rows grouped by episodes, though this may occasionally be the case (however knowledge of such a structure resides  
+    with the user as RLlib can't easily infer it automatically). While it's possible to concatenate :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
+    chunks, this can't be done with fragmented chunks.
```
```suggestion
    with the user as RLlib can't easily infer it automatically). While it's possible to concatenate consecutive :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
    chunks, this can't be done with chunks arriving in some scrambled order.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
+
+If you require full trajectories you can transform your tabular data into :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` objects and store these in Parquet format. The next example shows 
+how to do this. 
+First, you store experiences of the preceding trained expert policy in tabular format:
```
```suggestion
First, you store experiences of the preceding trained expert policy in tabular format (note the `output_write_episodes=False` setting below to activate tabular data output):
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2083 @@
     algo = config.build()
-    for _ in range(100):
-        algo.train()
+    # Load now the PPO-trained `RLModule` to use in recording.
```
```suggestion
    # Load the PPO-trained `RLModule` to use in recording.
```

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+    # Show the transformed batch.
+    print(f"Batch: {batch}")
+
+In the code example provided, you define a custom :py:class:`~ray.rllib.offline.offline_data.OfflineData` class to handle the reading and preprocessing of image data, converting it from a binary encoding format into `numpy` arrays. Additionally, you implement a custom :py:class:`~ray.rllib.offline.offline_prelearner.OfflinePreLearner` to process this data further, transforming it into a learner-r ...
```
Putting the above code into its own examples script allows us to describe the process some more in the header docstring of the example script. Then, it would be enough to just say here:

```
See here for a detailed example of how to write your own image preprocessing pipeline ...
```

simonsays1980 replied:
Done

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+----------
+To effectively monitor your offline data pipeline, leverage :ref:`Ray Data's built-in monitoring capacities <monitoring-your-workload>`. Focus on ensuring that all stages of your offline data streaming pipeline are actively processing data. Additionally, keep an eye on the Learner instance, particularly the `learner_update_timer`, which should maintain low values - around `0.02` for small models  ...
+
+.. note:: RLlib doesn't include :ref:`Ray Data <data>`  metrics in its results or display them in `Tensorboard` through :ref:`Ray Tune <tune-main>`'s :py:class:`~ray.tune.logger.tensorboardx.TBXLoggerCallback`. It's strongly recommended to enable the :ref:`Ray Dashboard <observability-getting-started>`, accessible at `127.0.0.1:8265`, for comprehensive monitoring and insights.
```
Interesting. Is there any way to funnel these metrics from data into RLlib's metrics loggers somehow? Not top-prio at this point, but would be super nice to have.

simonsays1980 replied:
I doubt that it will be smooth - instead I guess it will produce much overhead. In addition the metrics are  nested such that they can be optimally shown in the Ray Dashboard. 

I am also unsure at this point if these metrics should be in our TensorBoard as they are more system performance metrics than RL metrics. In larger companies these metrics also get checked by System Administrators/Hardware Engineers while the RL metrics get checked by RL engineers. But I am open for discussion.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+
+You can configure experience input for an agent using the following options:
+
+.. code-block:: python
```
We should not repeat the docstring of this method here. Can we just reference the `AlgorithmConfig.offline_data` method somehow and enhance the existing comments/docstrings by what you wrote here below (if that's not the case yet)?

simonsays1980 replied:
We can and we should. It makes it easier to maintain. So, what should we show here?

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
 Output API
 ----------
 
 You can configure experience output for an agent using the following options:
```
nit: let's write 1-2 more sentences here on what exactly gets configured in this method and why?
* Use this method for configuring how any algorithm in RLlib (online- or offline) writes its collected experiences to a data storage.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+Input API
+---------
+
+You can configure experience input for an agent using the following options:
```
nit: let's write 1-2 more sentences here on what exactly gets configured in this method and why?

* Use this method for configuring from where and how an offline RL algorithm reads its data input.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
-.. autoclass:: ray.rllib.offline.OutputWriter
-    :members:
-    :noindex:
+    def offline_data(
```
Ah, I see. You wanted to split this into in- and output. Ok.

Hmm, if it's that distinguishable, maybe we should separate this into 2 methods?

Ok, let's discuss offline, before we make a move here.

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+
+Using Old API Stack ``SampleBatch`` Recordings
+----------------------------------------------
+If you have expert data previously recorded using RLlib's older API stack, it can be seamlessly utilized in the new Offline RL API by setting ``input_read_sample_batches=True``. Alternatively, 
```
```suggestion
If you have expert data previously recorded using RLlib's old API stack, it can be seamlessly utilized in the new stack's Offline RL API by setting ``input_read_sample_batches=True``. Alternatively, 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
 
-You can configure experience input for an agent using the following options:
+.. note:: Typically, increasing the number of CPUs is sufficient for performance tuning in the post-processing stage of your pipeline. GPUs are only needed in specialized cases, such as in customized pipelines. For example, RLlib’s :py:class:`~ray.rllib.algorithms.marwil.marwil.MARWIL` implementation uses the 
+    :py:class:`~ray.rllib.connectors.learner.general_advantage_estimation.GeneralAdvantageEstimation` connector in its :py:class:`~ray.rllib.connectors.connector_pipeline_v2.ConnectorPipelineV2` to apply `General Advantage Estimation <https://arxiv.org/abs/1506.02438>`_ on experience batches. In these calculations the value model of the algorithm's 
```
```suggestion
    :py:class:`~ray.rllib.connectors.learner.general_advantage_estimation.GeneralAdvantageEstimation` connector in its :py:class:`~ray.rllib.connectors.connector_pipeline_v2.ConnectorPipelineV2` to apply `General Advantage Estimation <https://arxiv.org/abs/1506.02438>`_ on experience batches. In these calculations, the value model of the algorithm's 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+    :py:class:`~ray.rllib.connectors.learner.general_advantage_estimation.GeneralAdvantageEstimation` connector in its :py:class:`~ray.rllib.connectors.connector_pipeline_v2.ConnectorPipelineV2` to apply `General Advantage Estimation <https://arxiv.org/abs/1506.02438>`_ on experience batches. In these calculations the value model of the algorithm's 
+    :py:class:`~ray.rllib.core.rl_module.RLModule` is applied, which you can accelerate by running on a GPU.
+
+As an example, to provide each of your ``4`` :py:class:`~ray.rllib.offline.offline_prelearner.OfflinePreLearner` in the **Post-Processing (PreLearner)** ``2`` cpus you could use the following syntax:
```
```suggestion
As an example, to provide each of your ``4`` :py:class:`~ray.rllib.offline.offline_prelearner.OfflinePreLearner` in the **Post-Processing (PreLearner)** ``2`` CPUs you can use the following syntax:
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+    )
+
+.. warning:: Don't override the ``batch_size`` in RLlib's ``map_batches_kwargs``. This usually leads to high performance degradations. Note, this ``batch_size`` differs from the `train_batch_size_per_learner`: the former specifies the batch size in transformations of
+    the streaming pipeline, while the latter defines the batch size used for training within each :py:class:`~ray.rllib.core.learner.learner.Learner`.
```
```suggestion
    the streaming pipeline, while the latter defines the batch size used for training within each :py:class:`~ray.rllib.core.learner.learner.Learner` (the batch size of the actual model forward- and backward passes performed for training).
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
-You can create a custom input reader like the following:
+    from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+
+    AlgorithmConfig()
```
nit: Let's format these examples with the standard config format and create a `config` variable:

```
config = (
    AlgorithmConfig
    .training(
        ...
    )
    .offline_data(
        ...
    )
)
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
-    class CustomInputReader(InputReader):
-        def __init__(self, ioctx: IOContext): ...
-        def next(self): ...
+    AlgorithmConfig()
```
nit: same here: change format to above suggestion and add `config` variable.

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+    # Set the stopping metric to be the evaluation episode return mean.
+    metric = f"{EVALUATION_RESULTS}/{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}"
+
+    # Configure the Ray Tune.
```
```suggestion
    # Configure Ray Tune.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+        },
+    )
+
+.. tip:: To choose an adequate `input_read_batch_size` take a look at the length of your recorded episodes. In some cases each single episode is long enough to fullfill the `train_batch_size_per_learner` and you could choose a `input_read_batch_size` of ``1``. Most times it's not and you need to consider how many episodes should be buffered to balance
```
```suggestion
.. tip:: To choose an adequate `input_read_batch_size` take a look at the length of your recorded episodes. In some cases each single episode is long enough to fulfill the `train_batch_size_per_learner` and you could choose a `input_read_batch_size` of ``1``. Most times it's not and you need to consider how many episodes should be buffered to balance
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+Actor Pool Size
+***************
+
+RLlib supports scaling :py:class:`~ray.rllib.core.learner.learner.Learner` instances through the parameter `num_learners`. When this vcalue exceeds ``1``, scaling is implemented internally using a :py:class:`~ray.train._internals.backend_executor_BackendExecutor`. This executor spawns your specified 
```
```suggestion
RLlib supports scaling :py:class:`~ray.rllib.core.learner.learner.Learner` instances through the parameter `num_learners`. When this value is ``0``, RLlib uses a Learner instance in the local process, whereas for values ``>0``, RLlib scales out using a :py:class:`~ray.train._internals.backend_executor_BackendExecutor`. This executor spawns your specified 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+
+    from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+
+    AlgorithmConfig()
+    .learners(
+        num_learners=4,
+    )
```
```suggestion
    config = (
        AlgorithmConfig()
        .learners(num_learners=4)
    )
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+Allocated Resources
+~~~~~~~~~~~~~~~~~~~
+Just as with the Post-Processing (Pre-Learner) layer, allocating additional resources can help address slow training processes. In this case, the primary resource to leverage is the GPU, as training involves forward and backward passes through the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`, which GPUs can accelerate significantly. If your training is 
+already utilize GPUs and performance remains an issue, consider scaling up by either adding more GPUs to each :py:class:`~ray.rllib.core.learner.learner.Learner` to increase GPU memory and computational capacity, or by adding additional :py:class:`~ray.rllib.core.learner.learner.Learner` workers to further distribute the workload. Additionally, ensure that data 
```
```suggestion
already utilizes GPUs and performance still remains an issue, consider scaling up by either adding more GPUs to each :py:class:`~ray.rllib.core.learner.learner.Learner` to increase GPU memory and computational capacity (set `config.learners(num_gpus_per_learner=...)`), or by adding additional :py:class:`~ray.rllib.core.learner.learner.Learner` workers to further distribute the workload (by setting `config.learners(num_learners=...)`). Additionally, ensure that data 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+
+    from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+
+    AlgorithmConfig()
+    .learners(
+        num_learners=4,
+        num_gpus_per_learner=2,
+    )
```
```suggestion
    config = (
        AlgorithmConfig()
        .learners(num_learners=4, num_gpus_per_learner=2)
    )
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2081 @@
+        num_gpus_per_learner=2,
+    )
+
+.. tip::If you experience backpressure in the **Post-Processing (Pre-Learner)** stage of your pipeline, consider enabling GPU training before scaling up your :py:class:`~ray.rllib.core.learner.learner.Learner` instances.
```
```suggestion
.. tip::If you experience backpressure in the **Post-Processing (Pre-Learner)** stage of your pipeline, consider enabling GPU training before scaling up the number of your :py:class:`~ray.rllib.core.learner.learner.Learner` instances.
```

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
+
+Just as online training can be scaled, offline recording I/O throughput can also be increased by configuring the number of RLlib env-runners. Use the ``num_env_runners`` setting to scale recording during training or ``evaluation_num_env_runners``
+for scaling during evaluation-only recording. Each worker operates independently, writing experiences in parallel, enabling linear scaling of I/O throughput for write operations. Within each :py:class:`~ray.rllib.offline.offline_env_runner.OfflineSingleAgentEnvRunner`, episodes 
+are sampled (environments are usually vectorized) and serialized before being written to disk.
```
```suggestion
are sampled and serialized before being written to disk.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
 ---------------
 
-RLlib's offline dataset APIs enable working with experiences read from offline storage (e.g., disk, cloud storage, streaming systems, HDFS). For example, you might want to read experiences saved from previous training runs, or gathered from policies deployed in `web applications <https://arxiv.org/abs/1811.00260>`__. You can also log new agent experiences produced during online training for futur ...
+RLlib's offline RL API enables you to work with experiences read from offline storage (for example, disk, cloud storage,
+streaming systems, Hadoop Distributed File System/HDFS). For example, you might want to read experiences saved from previous training runs, collected
+from experts, or gathered from policies deployed in `web applications <https://arxiv.org/abs/1811.00260>`__. You can 
+also log new agent experiences produced during online training for future use.
```
```suggestion
RLlib's offline RL API enables you to work with experiences read from offline storage, such as files on disk, cloud storage, streaming systems, or Hadoop Distributed File System/HDFS. For example, you might want to read experiences saved from previous training runs, collected
from experts, or gathered from policies deployed in `web applications <https://arxiv.org/abs/1811.00260>`__. You can 
also log new agent experiences produced during online training for future use.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
 
-RLlib represents trajectory sequences (i.e., ``(s, a, r, s', ...)`` tuples) with `SampleBatch <https://github.com/ray-project/ray/blob/master/rllib/policy/sample_batch.py>`__ objects. Using a batch format enables efficient encoding and compression of experiences. During online training, RLlib uses `policy evaluation <rllib-concepts.html#policy-evaluation>`__ actors to generate batches of experien ...
+RLlib represents trajectory sequences (for example, ``(s, a, r, s', ...)`` tuples) with :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode`
+objects (multi-agent offline training is currently not supported). Using this episode format 
+allows for efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getter methods.
```
```suggestion
objects (multi-agent offline training is currently not supported). This episode format 
allows for efficient encoding and compression of experiences, rewriting trajectories, and user-friendly data access through getter methods.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
 
-.. code-block:: bash
+.. note:: RLlib's new API stack incorporates principles that support standalone applications. Consequently, the 
+    :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` class is available for use independently of RLlib. To enable faster 
```
```suggestion
    :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` class is usable outside of an RLlib context. To enable faster 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
 
-Example: Training on previously saved experiences
--------------------------------------------------
+You can store experiences either directly in RLlib's episode format or in table (columns) 
+format. You should use the episode format when 
```
```suggestion
You can store experiences either directly in RLlib's episode format or in tabular format (column data). You should use episodes when 
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
 
-In this example, we will save batches of experiences generated during online training to disk, and then leverage this saved data to train a policy offline using DQN. First, we run a simple policy gradient algorithm for 100k steps with ``"output": "/tmp/cartpole-out"`` to tell RLlib to write simulation outputs to the ``/tmp/cartpole-out`` directory.
+#. You need to read the data easily with other data tools or ML libraries.
+#. You don't need full (usually complete) trajectories in training.
```
```suggestion
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
-.. code-block:: bash
+.. note:: RLlib's new API stack incorporates principles that support standalone applications. Consequently, the 
+    :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` class is available for use independently of RLlib. To enable faster 
+    access through external data tools (for example, for data transformations), it's recommended to use the table record format in these scenarios.
```
```suggestion
    access through external data tools (for example, for data transformations), it's recommended to use the table record format.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
-The experiences will be saved in compressed JSON batch format:
+.. hint:: During the transition phase from old- to new API stack you can use the new offline RL API also with your 
+    :py:class:`~ray.rllib.policy.sample_batch.SampleBatch` data recorded with the old API stack. To enable this feature set 
+    ``input_read_sample_batches=True``.
```
```suggestion
    ``config.offline_data(input_read_sample_batches=True)``.
```

sven1977 commented on doc/source/rllib/rllib-offline.rst:
```diff
@@ -5,437 +5,2080 @@
+Example: Training an Expert Policy
+----------------------------------
+In this example you train a PPO agent on the ``CartPole-v1`` environment until it reaches an episode mean return of ``450.0``. You checkpoint 
+this agent and later use its policy to record expert data to local disk.
```
```suggestion
this agent and then use its policy to record expert data to local disk.
```

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

simonsays1980 committed:
Wrapped 'AlgorithmConfig' in brackets.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Removed code from RLlib's offline docs and instead used 'literalinclude' to link to example code.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 committed:
Merge branch 'master' into offline-rl-docs-overhaul

stephanie-wang committed:
[core][compiled graphs] Support arbitrary torch.dtypes when passing through shared memory (#48957)

Some torch.dtypes don't have a numpy equivalent. We use numpy to store
the tensor data zero-copy in the object store. To support these tensors,
we first view the array with a common dtype (uint8), and then view as a
np array. During deserialization, we use another view back to the
original dtype.

Closes #48141.

---------

Signed-off-by: Stephanie Wang <smwang@cs.washington.edu>

ravi-dalal committed:
[Tune] update optuna_search to allow users to configure optuna storage (#48547)

The OptunaSearch currently uses in-memory storage and does not provide a
way to configure any other storages (e.g. database). This update add a
configurable parameter to OptunaSearch that can be set to a valid Optuna
Storage.

Signed-off-by: Ravi Dalal <ravidalal@google.com>
Signed-off-by: Ravi Dalal <12639199+ravi-dalal@users.noreply.github.com>
Signed-off-by: Hongpeng Guo <hpguo@anyscale.com>
Co-authored-by: Hongpeng Guo <hg5@illinois.edu>
Co-authored-by: matthewdeng <matthew.j.deng@gmail.com>
Co-authored-by: Hongpeng Guo <hpguo@anyscale.com>
Co-authored-by: Justin Yu <justinvyu@anyscale.com>

dentiny committed:
[core] Cache runtime env for core worker (#49140)

Signed-off-by: hjiang <dentinyhao@gmail.com>

sven1977 committed:
[RLlib] Fix bugs in `MultiAgentEpisode` class (getters, `env_t_to_agent_t` system, etc..). (#49191)

bveeramani committed:
[Data] Revise training benchmarks (#49171)

In an effort to have fewer but more useful release tests, this PR removes all of the training release tests except for the distributed Parquet training and chaos distributed Parquet training release tests. Notably, this PR removes variants for different input formats, and single-node training release tests.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

dayshah committed:
[core] Stabilize many_nodes release test (#49193)

Signed-off-by: dayshah <dhyey2019@gmail.com>

mattip committed:
[Core] Fix rpdb.py for windows and encoding failures, adjust tests (#48117)

Signed-off-by: Matti Picus <matti.picus@gmail.com>

Commit:
[Dashboard] Add instance variables to many default dashboard graphs (#49174)

Signed-off-by: Alan Guo <aguo@anyscale.com>

win5923 committed:
[Doc] Logging: Add Fluent Bit sidecar and Grafana Loki to Persist Ray logs (#49038)

Signed-off-by: win5923 <ken89@kimo.com>

coltwood93 committed:
[Dashboard] Display duration in milliseconds if under 1 second. (#49126)

Signed-off-by: Colton Woodruff <coltwood93@gmail.com>

simonsays1980 committed:
Added missing image to git-tracked files.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

sven1977 committed:
[RLlib; docs] Docs do-over (new API stack): Env pages vol 01. (#49165)

sven1977 committed:
[RLlib] Cleanup examples folder (new API stack) vol 31: Add hierarchical training example script. (#49127)

dentiny committed:
[core] Use single asio context along the IO path (#49172)

Signed-off-by: hjiang <dentinyhao@gmail.com>

matthewdeng committed:
[train] reference custom resources in ScalingConfig.resources_per_worker (#48708)

Make it clear that you can specify Ray Core custom resources here.

Signed-off-by: Matthew Deng <matt@anyscale.com>

sven1977 committed:
[CI; RLlib] Increase parallel example scripts processes from 5 to 6. (#49217)

Jay-ju committed:
[data] webdataset - expand JSON objects into individual samples (#48673)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this
solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: jukejian <jukejian@bytedance.com>
Co-authored-by: srinathk10 <68668616+srinathk10@users.noreply.github.com>

kevin85421 committed:
[core][observability] Improve observability when the core worker fails to read data from the Raylet socket (#49163)

Signed-off-by: kaihsun <kaihsun@anyscale.com>

Jay-ju committed:
[data] remove _split_list dead code (#47954)


Signed-off-by: jukejian <jukejian@bytedance.com>

raulchen committed:
[Data] support passing kwargs to map tasks. (#49208)

## Why are these changes needed?

This PR enables passing kwargs to map tasks, which will be accessible
via `TaskContext.kwargs`.

This is a prerequisite to fixing
https://github.com/ray-project/ray/issues/49207. And optimization rules
can use this API to pass additional arguments to the map tasks.

---------

Signed-off-by: Hao Chen <chenh1024@gmail.com>

raulchen committed:
[data] Add ExecutionCallback interface (#49205)

## Why are these changes needed?

Add an ExecutionCallback interface to allow hooking custom callback
logic into certain execution events. This can be useful for optimization
rules.

---------

Signed-off-by: Hao Chen <chenh1024@gmail.com>

zcin committed:
[serve] Add DEPLOY_FAILED deployment status (#49178)

## Why are these changes needed?

Add a `DEPLOY_FAILED` deployment status.
- If replicas fail to start or health checks fail _during_ deploy, the
deployment status transitions to `DEPLOY_FAILED`
- If any deployments are in `DEPLOY_FAILED`, the application is also in
`DEPLOY_FAILED`.

## Related issue number

https://github.com/ray-project/ray/issues/48654


---------

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>
Co-authored-by: Gene Der Su <gdsu@ucdavis.edu>

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2614,11 +2614,11 @@ class here, if you need some primer transformations specific to your
-                are passed into the read method without checking. If no arguments
-                are passed in the default argument
-                `{'override_num_blocks': max(num_learners * 2, 2)}` is used. Use these
+                are passed into the read method without checking. Use these
```
```suggestion
                are passed by RLlib into the read method without checking. Use these
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2636,8 +2637,8 @@ class here, if you need some primer transformations specific to your
                 unsure when to use the data or in which RL framework. The default is
-                to read column data, i.e. False. `input_read_episodes` and
-                `input_read_sample_batches` cannot be True at the same time. See
+                to read column data, i.e. `False`. `input_read_episodes` and
```
```suggestion
                to read column data, for example, `False`. `input_read_episodes`, and
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2636,8 +2637,8 @@ class here, if you need some primer transformations specific to your
-                to read column data, i.e. False. `input_read_episodes` and
-                `input_read_sample_batches` cannot be True at the same time. See
+                to read column data, i.e. `False`. `input_read_episodes` and
+                `input_read_sample_batches` cannot be `True` at the same time. See
```
```suggestion
                `input_read_sample_batches` can't be `True` at the same time. See
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2646,21 +2647,23 @@ class here, if you need some primer transformations specific to your
-                `MultiAgentEpisode`). The default is False. `input_read_episodes`
-                and `input_read_sample_batches` cannot be True at the same time.
+                RLlib's `EpisodeType` (i.e. `SingleAgentEpisode`). The default is
+                `False`. `input_read_episodes` and `input_read_sample_batches` cannot
```
```suggestion
                `False`. `input_read_episodes`, and `input_read_sample_batches` can't
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2646,21 +2647,23 @@ class here, if you need some primer transformations specific to your
-                `BatchType` (i.e. `SampleBatch` or `MultiAgentBatch`) or any other
-                data type that contains multiple timesteps in a single row of the
-                dataset. In such cases a single batch of size
+                `EpisodeType` (i.e. `SingleAgentEpisode`) or `SampleBatch`, or any
```
```suggestion
                `EpisodeType` (i.e., `SingleAgentEpisode`) or `SampleBatch`, or any
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2716,6 +2718,17 @@ class here, if you need some primer transformations specific to your
                 need to make some further transformations specific for your data or
                 loss. The default is None which uses the base `OfflinePreLearner`
                 defined in `ray.rllib.offline.offline_prelearner`.
+            prelearner_buffer_class: An optional `EpisodeReplayBuffer` class is
```
```suggestion
            prelearner_buffer_class: An optional `EpisodeReplayBuffer` class that RLlib
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2716,6 +2718,17 @@ class here, if you need some primer transformations specific to your
                 loss. The default is None which uses the base `OfflinePreLearner`
                 defined in `ray.rllib.offline.offline_prelearner`.
+            prelearner_buffer_class: An optional `EpisodeReplayBuffer` class is
+                used to buffer experiences when data is in `EpisodeType` or
```
```suggestion
                uses to buffer experiences when data is in `EpisodeType` or
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2716,6 +2718,17 @@ class here, if you need some primer transformations specific to your
+                RLlib's previous `SampleBatch` type format. In this case, a single
+                data row may contain multiple timesteps and the buffer serves two
+                purposes: (a) to store intermediate data in memory, and (b) to ensure
+                that exactly `train_batch_size_per_learner` experiences are sampled
```
```suggestion
                that RLlib samples exactly `train_batch_size_per_learner` experiences
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2716,6 +2718,17 @@ class here, if you need some primer transformations specific to your
+                that exactly `train_batch_size_per_learner` experiences are sampled
+                per batch. The default is RLlib's `EpisodeReplayBuffer`.
+            prelearner_buffer_kwargs: Optional keyword arguments for intializing the
+                `EpisodeReplayBuffer`. In most cases this is simply the `capacity`
```
```suggestion
                `EpisodeReplayBuffer`. In most cases this value is simply the `capacity`
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2716,6 +2718,17 @@ class here, if you need some primer transformations specific to your
+                per batch. The default is RLlib's `EpisodeReplayBuffer`.
+            prelearner_buffer_kwargs: Optional keyword arguments for intializing the
+                `EpisodeReplayBuffer`. In most cases this is simply the `capacity`
+                for the default buffer used (`EpisodeReplayBuffer`), but it may
```
```suggestion
                for the default buffer that RLlib uses (`EpisodeReplayBuffer`), but it may
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2727,6 +2740,7 @@ class here, if you need some primer transformations specific to your
                 during a single training iteration. If None, each learner runs a
                 complete epoch over its data block (the dataset is partitioned into
                 at least as many blocks as there are learners). The default is `None`.
+                This must be set to `1`, if a single (local) learner is used.
```
```suggestion
                This value must be set to `1`, if RLlib uses a single (local) learner.
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2779,6 +2793,12 @@ class here, if you need some primer transformations specific to your
                 given by `output_filesystem`. See `gcsfs.GCSFilesystem` for GCS,
                 `pyarrow.fs.S3FileSystem`, for S3, and `ablfs.AzureBlobFilesystem` for
                 ABS filesystem arguments.
+            output_write_episodes: If data should be recorded in RLlib's `EpisodeType`
```
```suggestion
            output_write_episodes: If RLlib should record data in its RLlib's `EpisodeType`
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2779,6 +2793,12 @@ class here, if you need some primer transformations specific to your
                 `pyarrow.fs.S3FileSystem`, for S3, and `ablfs.AzureBlobFilesystem` for
                 ABS filesystem arguments.
+            output_write_episodes: If data should be recorded in RLlib's `EpisodeType`
+                format (i.e. `SingleAgentEpisode` objects). Use this format, if you
```
```suggestion
                format (that is, `SingleAgentEpisode` objects). Use this format, if you
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2779,6 +2793,12 @@ class here, if you need some primer transformations specific to your
                 ABS filesystem arguments.
+            output_write_episodes: If data should be recorded in RLlib's `EpisodeType`
+                format (i.e. `SingleAgentEpisode` objects). Use this format, if you
+                need data to be ordered in time and directly grouped by episodes for
```
```suggestion
                need RLlib to order data in time and directly group by episodes for
```

angelinalg commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2779,6 +2793,12 @@ class here, if you need some primer transformations specific to your
+                format (i.e. `SingleAgentEpisode` objects). Use this format, if you
+                need data to be ordered in time and directly grouped by episodes for
+                example to train stateful modules or if you plan to use recordings
+                exclusively in RLlib. Otherwise data is recorded in tabular (columnar)
```
```suggestion
                exclusively in RLlib. Otherwise RLlib records data in tabular (columnar)
```

angelinalg reviewed (approved):
Sorry for the delay, @simonsays1980 and @sven1977. Thanks for letting me review. I'm mostly suggesting some style nits to rewrite passive voice sentences to improve clarity. Apologies if I used the incorrect subject.

jecsand838 committed:
[Data] Add read_clickhouse API to read ClickHouse Dataset (#49060)

Greetings from ElastiFlow!

This PR introduces a new ClickHouseDatasource connector for Ray, which
provides a convenient way to read data from ClickHouse into Ray
Datasets. The ClickHouseDatasource is particularly useful for users who
are working with large datasets stored in ClickHouse and want to
leverage Ray's distributed computing capabilities for AI and ML
use-cases. We found this functionality useful while evaluating ML
technologies and wanted to contribute this back.

Key Features and Benefits:
1. **Seamless Integration**: The ClickHouseDatasource allows for
seamless integration of ClickHouse data into Ray workflows, enabling
users to easily access their data and apply Ray's powerful parallel
computation.
2. **Custom Query Support**: Users can specify custom columns, and
orderings, allowing for flexible query generation directly from the Ray
interface, which helps in reading only the necessary data, thereby
improving performance.
3. **User-Friendly API**: The connector abstracts the complexity of
setting up and querying ClickHouse, providing a simple API that allows
users to focus on data analysis rather than data extraction.

Tested locally with a ClickHouse table containing ~12m records.

<img width="1340" alt="Screenshot 2024-11-20 at 3 52 42 AM"
src="https://github.com/user-attachments/assets/2421e48a-7169-4a9e-bb4d-b6b96f7e502b">

PLEASE NOTE: This PR is a continuation of
https://github.com/ray-project/ray/pull/48817, which was closed without
merging.

---------

Signed-off-by: Connor Sanders <connor@elastiflow.com>
Co-authored-by: Alexey Kudinkin <alexey.kudinkin@gmail.com>

zcin committed:
[serve] Add missing DEPLOY_FAILED state transitions (#49225)

## Why are these changes needed?

Allow deployment to transition out of `DEPLOY_FAILED` into `UPDATING`.


Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

dentiny committed:
[core] Cleanup shared pointer for ref counter in memory store (#49212)

Signed-off-by: hjiang <hjiang@anyscale.com>

bveeramani committed:
[Data] Add release test for TPCH Q1 (#49197)

To make sure Ray Data can performantly handle complex data processing
workloads, this PR adds a TPCH benchmark to our release tests.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

sven1977 committed:
Apply suggestions from code review

Co-authored-by: angelinalg <122562471+angelinalg@users.noreply.github.com>
Signed-off-by: Sven Mika <sven@anyscale.io>

sven1977 reviewed (approved):
LGTM now, thanks so much for this great PR @simonsays1980  and your review @angelinalg !

simonsays1980 committed:
Fixed some linter errors.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed two class reference bugs.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

can-anyscale commented:
doc build errors


```
[2024-12-13T12:16:12Z] /ray/doc/source/rllib/rllib-offline.rst:6: WARNING: Duplicate explicit target name: "here".
--
  | [2024-12-13T12:16:12Z] /ray/doc/source/rllib/rllib-offline.rst:6: WARNING: Duplicate explicit target name: "here".
  | [2024-12-13T12:16:12Z] looking for now-outdated files... none found
  | [2024-12-13T12:16:12Z] pickling environment... done
  | [2024-12-13T12:16:12Z] checking consistency... done
  | [2024-12-13T12:16:12Z] preparing documents... done
  | copying downloadable files... [100%] /python/ray/tune/examples/tune-default.yaml
  | [2024-12-13T12:16:12Z] copying static files... done
  | [2024-12-13T12:16:12Z] copying extra files... done
  | [2024-12-13T12:16:12Z] done
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:237: WARNING: py:meth reference target not found: ray.rllib.offline.offline_env_runner.OfflineSingleAgentEnvRunner.sample
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:351: WARNING: py:data reference target not found: ray.rllib.offline.offline_prelearner.SCHEMA
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:1326: WARNING: py:meth reference target not found: ray.rllib.offline.offline_prelearner.OfflinePreLearner._map_to_episodes
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:1443: WARNING: py:meth reference target not found: ray.rllib.offline.offline_prelearner.OfflinePreLearner.__call__
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:1700: WARNING: py:meth reference target not found: ray.rllib.offline.offline_data.OfflineData.__init__
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:1701: WARNING: py:meth reference target not found: ray.rllib.offline.offline_data.OfflineData.sample
  | [2024-12-13T12:16:36Z] /ray/doc/source/rllib/rllib-offline.rst:1703: WARNING: py:meth reference target not found: ray.rllib.offline.offline_data.OfflineData.__init__
```

simonsays1980 committed:
Added Offline RL API reference to include new modules.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merged master and resolved conflicts.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed a small bug in automsummaries.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
FIxed string error in 'OfflinePreLearner__call__'.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed reference error for offline API reference. The old stack is still in there.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added offline.rst to index.rst b/c it was not referenced anywhere, yet, which produces an error in the docs tests.§

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed some more reference bugs due to missing automsummaries.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
CHanged autodoc for SCHEMA by adding special docstring for data attribute.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

sven1977 committed:
Merge branch 'master' of https://github.com/ray-project/ray into offline-rl-docs-overhaul

sven1977 committed:
wip

sven1977 committed:
wip

angelinalg committed:
[docstring] adding link to runtime_env API ref to instances of ray_remote_args (#48955)

Per request from @GokuMohandas, add link to runtime_env API ref for all
commands with the `ray_remote_args` parameter to improve usability.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---------

Signed-off-by: angelinalg <angelina@anyscale.com>
Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

win5923 committed:
[Doc] Fix incorrect URL in `persist-kuberay-custom-resource-logs.md` (#49323)

Signed-off-by: win5923 <ken89@kimo.com>

dentiny committed:
[core] Minor update on task manager (#49272)

Signed-off-by: hjiang <hjiang@anyscale.com>

dentiny committed:
[core] Fix cpp inclusion check (#49301)

Signed-off-by: hjiang <dentinyhao@gmail.com>

dentiny committed:
[core] Remove unused function arg (#49331)

Signed-off-by: dentiny <dentinyhao@gmail.com>

simonsays1980 committed:
Removed recording examples from CI tests b/c it times out.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

simonsays1980 committed:
Added a line to code-block as this was failing in doc tests.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

simonsays1980 committed:
Merged master and resolved conflicts.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Removed another test code block as test was timing out.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Fixed an import error in example and ran vale.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Changed 'BCTorchRLModule into DefaultBCTorchRLModule, which was causing a bug.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul

simonsays1980 committed:
Fixed a bug in the text example.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-docs-overhaul
